### PR TITLE
Essence damage

### PR DESCRIPTION
**Pull request by JFranklin3000**
Closes [[issue URL](https://github.com/WookieeMatt/Essence20/issues/553)]

##### In this PR
- Migrate essences to have value and max values
- Added a field on the essence to represent 

##### Testing
- Validate user migration finishes successfully.

### CODE REVIEW

File: module/data/actor/templates/character.mjs [hunk 1]
```diff
@@ -27,11 +34,40 @@ export const character = () => ({
...
+      source.essences.strength.value = strength;
+      source.essences.speed.max = speed;
+      source.essences.speed.value = speed;
+      source.essences.smarts.max = smarts;
+      source.essences.smarts.value = smarts;
+      source.essences.social.max = social;
+      source.essences.social.value = social;
+    }
+  }
+
```
Comment (phildominguez) [inline module/data/actor/templates/character.mjs:72]:
```suggestion
```

File: module/migration.mjs [hunk 2]
```diff
@@ -196,6 +196,7 @@ export const migrateActorData = async function(actor, compendiumActor) {
     }
   }
 
+
```
Comment (phildominguez) [inline module/migration.mjs:199]:
```suggestion
```

File: module/sheet-handlers/listener-misc-handler.mjs [hunk 3]
```diff
@@ -122,6 +122,16 @@ export async function onRest(actorSheet) {
     }
   }
 
+  for (const essence of Object.keys(actor.system.essences)) {
```
Comment (phildominguez) [inline module/sheet-handlers/listener-misc-handler.mjs:125]:
```suggestion
  // Recovering Essence damage
  for (const essence of Object.keys(actor.system.essences)) {
```

File: module/sheet-handlers/listener-misc-handler.mjs [hunk 4]
```diff
@@ -122,6 +122,16 @@ export async function onRest(actorSheet) {
...
+    if (actor.system.essences[essence].value < actor.system.essences[essence].max) {
+      const essenceString = `system.essences.${essence}.value`;
+      const essenceRestore = actor.system.essences[essence].value + 1;
+      await actor.update({
+        [essenceString]: essenceRestore,
+      });
```
Comment (phildominguez) [inline module/sheet-handlers/listener-misc-handler.mjs:131]: I would put a UI notification saying that the essence was recovered, like the other messages

**Review by phildominguez — CHANGES_REQUESTED** [review]

File: module/data/actor/companion.mjs [hunk 5]
```diff
@@ -31,3 +31,5 @@ export class CompanionActorData extends foundry.abstract.TypeDataModel {
     };
   }
 }
+
+
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply module/data/actor/companion.mjs]: reverted

File: module/sheet-handlers/listener-misc-handler.mjs (hunk 4 above)
Comment (JFranklin3000) [reply module/sheet-handlers/listener-misc-handler.mjs:132]: UI notification added.

File: lang/en.json [hunk 6]
```diff
@@ -405,6 +405,7 @@
     "RestComplete": "Rest complete!",
     "RestEnergonReset": "{energon} Energon reset.",
     "RestEnergonRestored": "Energon restored by {energonRestore}.",
+    "RestEssenceRestored": "{essence} essence restored to {essenceRestore}",
```
Comment (phildominguez) [inline lang/en.json:408]:
```suggestion
    "RestEssenceRestored": "{essence} Essence restored to {essenceRestore}",
```

**Review by phildominguez — CHANGES_REQUESTED** [review]

**Review by phildominguez — APPROVED** [review]

**Pull request merged**